### PR TITLE
[HOP] Mutation and alias rework

### PR DESCRIPTION
**Pull request by bohnstingl**
This PR reworks the way the input mutations and various aliases are checked

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @ydwu4

### CODE REVIEW

File: test/functorch/test_control_flow.py [hunk 1]
```diff
@@ -5938,7 +5945,8 @@ def f(xs, y):
         example_inputs = (torch.ones(3, 2, 4), torch.ones(4))
         functional_f = torch.func.functionalize(f)
         with self.assertRaisesRegex(
-            UnsupportedAliasMutationException, "torch.map is mutating the input!"
+            UnsupportedAliasMutationException,
+            "A branch or combine_fn might be modifying the input!.*",
```
Comment (ydwu4) [inline test/functorch/test_control_flow.py:5949]: The error message for hops can be different. We need to make sure the error messsages makes sense.

File: torch/_higher_order_ops/utils.py [hunk 2]
```diff
@@ -191,6 +210,40 @@ def _maybe_fake_tracing(fn, inputs: List[Any], pre_dispatch):
...
+        # functionalized
+        return True
+    except Exception as e:
+        raise e
+
+    return _detect_input_mutation(gm), _detect_alias(gm)
+
+
+def check_input_mutation_and_alias(combine_fn, sample_inputs, pre_dispatch=False):
+    _, has_branch_alias = check_input_mutation(
```
Comment (ydwu4) [inline torch/_higher_order_ops/utils.py:227]: why not call has_potential_input_mutation_or_alias directly?  "check_input_mutation" no need to return has_branch_alias anymore, which is pretty weird.

File: test/functorch/test_control_flow.py (hunk 1 above)
Comment (bohnstingl) [reply test/functorch/test_control_flow.py]: I reworked the error messages

File: torch/_dynamo/graph_deduplication.py [hunk 3]
```diff
@@ -120,6 +120,7 @@ def _replace_region_with_subgraph(
     invoke_args = (get_subgraph_node, subgraph_name, tuple(sub_args))
     fake_inputs = [node.meta["example_value"] for node in sub_args]
 
+    # TODO: We don't care here about any output-output aliasing?
     if has_potential_input_alias_or_mutation(sub_gm, fake_inputs):
         log.debug(
             "NYI: Failed to substitute region %s due to input alias or mutation",
```
[the remark this answers is not in the document]
Comment (bohnstingl) [reply torch/_dynamo/graph_deduplication.py]: I revised

File: torch/_dynamo/variables/higher_order_ops.py [hunk 4]
```diff
@@ -2957,6 +2957,7 @@ def call_function(
             for node in body_gmod.graph.nodes
             if node.op == "placeholder"
         ]
+        # TODO: We don't care here about any output-output aliasing?
         if has_potential_input_alias_or_mutation(body_gmod, fake_inputs):
             raise RuntimeError(
                 f"{self.value._name} where the inputs are mutated or the "
```
[the remark this answers is not in the document]
Comment (bohnstingl) [reply torch/_dynamo/variables/higher_order_ops.py]: I revised

File: torch/_higher_order_ops/utils.py (hunk 2 above)
Comment (bohnstingl) [reply torch/_higher_order_ops/utils.py]: I reworked that check and isolated the aliases and mutations such that it should be easier in a follow-up PR to determine which node is triggering the checks

**Comment by bohnstingl** [comment]
@ydwu4 I reworked the mutation and alias checks. I moved the checks into dynamo for scan, associative_scan, while_loop and cond. For map I also included the new check, but since it does not yet use backend='eager', I did not move the check to dynamo. Pleas let me know what you think

File: torch/_dynamo/variables/higher_order_ops.py [hunk 5]
```diff
@@ -127,6 +126,55 @@ def _unwrap_var(var):
 
     return check_meta_consistency(unwrapped1, unwrapped2, lhs_name, rhs_name)
 
+def flatten_nested_lists(li):
```
Comment (ydwu4) [inline torch/_dynamo/variables/higher_order_ops.py:129]: why not pytree.tree_flatten?

File: torch/_dynamo/variables/higher_order_ops.py [hunk 6]
```diff
@@ -127,6 +126,55 @@ def _unwrap_var(var):
...
+            if hasattr(inp, 'node'):
+                val = inp.node.meta["example_value"]
+                if isinstance(val, torch.Tensor):
+                    if torch._C._functorch.is_batchedtensor(val) or torch._C._functorch.is_functionaltensor(val):
+                        # This case is for batched or functional tensors
+                        while torch._C._functorch.is_batchedtensor(val) or torch._C._functorch.is_functionaltensor(val):
+                            val = torch._C._functorch.get_unwrapped(val)
+                        inputs_fake.append(val)
+                    else:
+                        # This is the standard case of a TensorVariable
```
Comment (ydwu4) [inline torch/_dynamo/variables/higher_order_ops.py:158]: can assert it's FakeTensor?

File: torch/_dynamo/variables/higher_order_ops.py [hunk 7]
```diff
@@ -127,6 +126,55 @@ def _unwrap_var(var):
...
+                        inputs_fake.append(val)
+                else:
+                    # This case is for SymInts and other non-Tensor elements
+                    inputs_fake.append(val)
+            else:
+                # This case is for ints
```
Comment (ydwu4) [inline torch/_dynamo/variables/higher_order_ops.py:164]: can assert they're ints?

File: torch/_higher_order_ops/flex_attention.py [hunk 8]
```diff
@@ -7,13 +7,12 @@
 from torch import Tensor
 from torch._C import DispatchKey
 from torch._higher_order_ops.utils import (
-    _has_potential_branch_input_mutation,
     _maybe_reenter_make_fx,
     autograd_not_implemented,
+    # check_input_mutation,
```
Comment (ydwu4) [inline torch/_higher_order_ops/flex_attention.py:12]: remove?

File: torch/_higher_order_ops/hints_wrap.py [hunk 9]
```diff
@@ -3,12 +3,10 @@
 import torch.utils._pytree as pytree
 from torch._C import DispatchKey
 from torch._higher_order_ops.utils import (
-    _has_potential_branch_input_alias,
-    _has_potential_branch_input_mutation,
     autograd_not_implemented,
+    # check_input_mutation_and_alias,
```
Comment (ydwu4) [inline torch/_higher_order_ops/hints_wrap.py:7]: remove?

File: torch/_higher_order_ops/scan.py [hunk 10]
```diff
@@ -442,27 +439,6 @@ def scan_functionalize(ctx, combine_fn, init, xs, additional_inputs):
     unwrapped_additional_inputs = ctx.unwrap_tensors(additional_inputs)
     with ctx.redispatch_to_next():
         functional_combine_fn = ctx.functionalize(combine_fn)
```
Comment (ydwu4) [inline torch/_higher_order_ops/scan.py:441]: we might still want to keep the check in functionalization key. In case someone is using the hop directly, which could bypass the dynamo checks.

File: torch/_higher_order_ops/utils.py [hunk 11]
```diff
@@ -201,43 +173,26 @@ def has_potential_input_alias_or_mutation(gm, inputs, pre_dispatch=False):
...
+        inp_mutation,
+        inp_inp_alias_map,
+        inp_out_alias_map,
+        out_out_alias_map,
+    ) = check_input_alias_and_mutation(gm, example_inputs)
+    return inp_mutation, (inp_inp_alias_map,
+                          inp_out_alias_map,
+                          out_out_alias_map)
+
+def has_potential_input_mutation_or_alias(gm, inputs, pre_dispatch=False):
```
Comment (ydwu4) [inline torch/_higher_order_ops/utils.py:189]: nit: Probably we should still name this as has_potential_input_alias_or_mutation?  The name change seems unnecessary

**Review by ydwu4 — COMMENTED** [review]
Looks good overall. Left a few minor comments.

File: torch/_dynamo/variables/higher_order_ops.py (hunk 5 above)
Comment (bohnstingl) [reply torch/_dynamo/variables/higher_order_ops.py:129]: Sure, I changed that.

File: torch/_dynamo/variables/higher_order_ops.py (hunk 7 above)
Comment (bohnstingl) [reply torch/_dynamo/variables/higher_order_ops.py:164]: Done.

I was wondering though in general whether we could somehow improve this? Are you fine with we currently collect the fake inputs?

File: torch/_higher_order_ops/flex_attention.py (hunk 8 above)
Comment (bohnstingl) [reply torch/_higher_order_ops/flex_attention.py:12]: Sure, I updated the PR and now reworked pretty much all HOPs

File: torch/_higher_order_ops/hints_wrap.py (hunk 9 above)
Comment (bohnstingl) [reply torch/_higher_order_ops/hints_wrap.py:7]: See above.

File: torch/_higher_order_ops/scan.py (hunk 10 above)
Comment (bohnstingl) [reply torch/_higher_order_ops/scan.py:441]: Right, as discussed offline, I reintroduced the checks into the `functionalization` key as well. 
If the HOP uses the backend='eager', we now we have it in `functionalization` key and in `dynamo`.

File: torch/_higher_order_ops/utils.py (hunk 11 above)
Comment (bohnstingl) [reply torch/_higher_order_ops/utils.py:189]: I agree. I corrected it.

File: torch/_higher_order_ops/while_loop.py [hunk 12]
```diff
@@ -467,8 +467,8 @@ def while_loop_func(ctx, cond_fn, body_fn, carried_inputs, additional_inputs):
         functional_body_fn = ctx.functionalize(_maybe_run_with_interpreter(body_fn))
         pre_dispatch = hasattr(ctx, "mode") and ctx.mode.pre_dispatch
         for fn, fn_name in [
-            (cond_fn, "cond_fn"),
-            (body_fn, "body_fn"),
+            (functional_cond_fn, "cond_fn"),
```
Comment (ydwu4) [inline torch/_higher_order_ops/while_loop.py:470]: why we need to change this?

Comment (bohnstingl) [reply torch/_higher_order_ops/while_loop.py]: I guess we don't have to use the functionalized variable. I reverted it and also for all the other HOPs we use the non functionalized variable for the mutation and alias checking.

**Comment by bohnstingl** [comment]
Currently we have only limited alias and mutation checks for `flex_attention`, i.e., we only check for input mutation and we also don't check either input alias or input mutation for `auto_functionalize`.

**Comment by bohnstingl** [comment]
There is a potential overlap with https://github.com/pytorch/pytorch/pull/148953 and it should be discussed what changes from here are relevant and should be kept and what is redundant.

**Comment by bohnstingl** [comment]
@ydwu4, @anijain2305: I updated this PR given the this [PR](https://github.com/pytorch/pytorch/pull/148953) and the discussion [here](https://github.com/pytorch/pytorch/issues/149989). What do you think about it?

File: test/functorch/test_control_flow.py [hunk 13]
```diff
@@ -888,8 +888,12 @@ def forward(self, arg0_1, arg1_1, arg2_1, arg3_1, arg4_1, arg5_1, arg6_1):
         )
 
     def test_cond_autograd_pytree_input(self):
+        # TODO: This is an unexpected behavior for cond
+        # Without this additional multiplication,
+        # the output of the backward graph would alias the
+        # inputs, as the gradients are just 1s and thus get optimized
         def true_fn(x):
-            return x["t"][0] + x["t"][1]["b"] * x["t"][2][0]
+            return (x["t"][0] * 2.0) + x["t"][1]["b"] * x["t"][2][0]
```
Comment (bohnstingl) [inline test/functorch/test_control_flow.py:896]: This may be a bit unexpected for to the user. Currently we don't allow aliases, including input-output aliases. This is problematic, because the gradients could be just 1s, in which case the gradients (arguments) from the upstream are just passed along, which then triggers the alias checks.

A naive solution would be to disable the input-output alias check, but I am not sure whether this causes problems? 
Is there another solution to this?

Comment (ydwu4) [reply test/functorch/test_control_flow.py:895]: A better way is to properly support it through auto_functionalized. This is still WIP though.

File: torch/_dynamo/output_graph.py [hunk 14]
```diff
@@ -2831,11 +2831,13 @@ def has_input_mutation(self):
...
 
     def has_aliasing(self):
+        from torch._higher_order_ops.utils import _collect_fake_inputs
+
         input_storages: dict[StorageWeakRef, torch.fx.Node] = dict()
 
         for node in self.graph.nodes:
             if node.op == "placeholder":
-                example_value = node.meta["example_value"]
+                example_value = _collect_fake_inputs([node])[0]
```
Comment (ydwu4) [inline torch/_dynamo/output_graph.py:2840]: what happens here? I don't expect it will error, will it?

File: torch/_dynamo/variables/higher_order_ops.py [hunk 15]
```diff
@@ -1336,6 +1340,8 @@ def create_unbacked_sym_node_var(tx) -> SymNodeVariable:
             source_target=self.value,
             set_subgraph_inputs="flatten_manual",
             should_flatten_outputs=True,
+            supports_input_mutation=False,
```
Comment (ydwu4) [inline torch/_dynamo/variables/higher_order_ops.py:1343]: Can put these two as class field? like BaseHOPVariable

File: torch/_higher_order_ops/utils.py [hunk 16]
```diff
@@ -699,17 +741,19 @@ def validate_subgraph_args_types(lifted_args: Union[tuple[Any, ...], list[Any]])
     ), f"{lifted_args} can only be of {allowed_types} but got {tuple(type(arg) for arg in lifted_args)}"
 
 
+# TODO: Return a more detailed information as to which node
+# causes a mutation or an alias. This may requires a per operator tensor version checking
 def check_input_alias_and_mutation(
```
Comment (ydwu4) [inline torch/_higher_order_ops/utils.py:746]: why we need to move mutated_inputs to the end?

**Review by ydwu4 — COMMENTED** [review]
Tests are failing. Added a few comments

File: torch/_higher_order_ops/utils.py (hunk 16 above)
Comment (bohnstingl) [reply torch/_higher_order_ops/utils.py:746]: Well, I just moved it in order to be consistent with the function name. It has alias and mutation and that's why I moved it towards the end. WDYT?

File: torch/_dynamo/variables/higher_order_ops.py (hunk 15 above)
Comment (bohnstingl) [reply torch/_dynamo/variables/higher_order_ops.py:1349]: Done.

File: torch/_dynamo/output_graph.py (hunk 14 above)
Comment (bohnstingl) [reply torch/_dynamo/output_graph.py:2840]: No, it does error out. The issue is that in some testcases. For example in, the issue is that the example value is a `BatchedTensorImpl`, for which `self.untyped_storage()` doesn't exist.

**Review by ydwu4 — APPROVED** [review]
The change looks good. Not sure why the test starts to fail.

**Comment by bohnstingl** [comment]
I think I found the issue. There was one occasion in the inductor where I missed the rearrangement of the return variable from alias_mutation. Fingers crossed for this time.
